constraints/src/main/java/gov/nasa/jpl/aerie/constraints/time/Spans.java

```diff
@@ -7,65 +7,102 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.function.Function;
 import java.util.function.Predicate;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 
 import gov.nasa.jpl.aerie.constraints.time.Interval.Inclusivity;
+import org.apache.commons.lang3.tuple.Pair;
 
 /**
  * A collection of intervals that can overlap.
  */
-public class Spans implements IntervalContainer<Spans>, Iterable<Interval> {
-  private final List<Interval> intervals;
+public class Spans implements IntervalContainer<Spans>, Iterable<Segment<Optional<Spans.Metadata>>> {
+  private final List<Segment<Optional<Metadata>>> intervals;
+
+  public record Metadata(long id){}
 
   public Spans() {
     this.intervals = new ArrayList<>();
   }
 
-  public Spans(final ArrayList<Interval> intervals) {
-    intervals.removeIf(Interval::isEmpty);
-    this.intervals = intervals;
+  public Spans(final List<Segment<Optional<Metadata>>> intervals) {
+    this.intervals = new ArrayList<>(intervals);
+  }
+
+  public Spans(final Spans spans){
+    this.intervals = new ArrayList<>(spans.intervals);
   }
 
   public Spans(final Iterable<Interval> iter) {
-    this.intervals = StreamSupport.stream(iter.spliterator(), false).filter($ -> !$.isEmpty()).toList();
+    this.intervals = new ArrayList<>();
+    StreamSupport.stream(iter.spliterator(), false).filter($ -> !$.isEmpty())
+                 .forEach(itv -> this.intervals.add(Segment.of(itv, Optional.empty())));
   }
 
   public Spans(final Interval... intervals) {
-    this.intervals = Arrays.stream(intervals).filter($ -> !$.isEmpty()).toList();
+    this.intervals = new ArrayList<>();
+    Arrays.stream(intervals).filter($ -> !$.isEmpty()).forEach(itv -> this.intervals.add(Segment.of(itv, Optional.empty())));
+  }
+
+  public Spans(final Segment<Optional<Metadata>>... intervals) {
```

Comment:
Because the varargs type is generic, and Java arrays do not handle generics well, Java gives us a warning unless we pinky promise that all the operations we're doing are type safe:

```suggestion
  @SafeVarargs
  public Spans(final Segment<Optional<Metadata>>... intervals) {
```
```
warning: [unchecked] unchecked generic array creation for varargs parameter of type Segment<Optional<Metadata>>[]
warning: [unchecked] Possible heap pollution from parameterized vararg type Segment<Optional<Metadata>>
```